avogadro/qtplugins/customelements/customelements.cpp

```diff
@@ -86,9 +99,123 @@ void CustomElements::reassign()
   }
 }
 
+bool CustomElements::openFile(const QString& fileName, Io::FileFormat* reader)
+{
+  if (fileName.isEmpty() || reader == nullptr) {
+    delete reader;
+    return false;
+  }
+
+  QString ident = QString::fromStdString(reader->identifier());
+
+  // Prepare the background thread to read in the selected file.
+  if (!m_fileReadThread)
+    m_fileReadThread = new QThread(qobject_cast<QWidget*>(parent()));
+  if (m_threadedReader)
+    m_threadedReader->deleteLater();
+  m_threadedReader = new BackgroundFileFormat(reader);
+  if (m_fileReadMolecule)
+    m_fileReadMolecule->deleteLater();
+  m_fileReadMolecule = new Molecule(qobject_cast<QWidget*>(parent()));
+  m_fileReadMolecule->setData("fileName", fileName.toLocal8Bit().data());
+  m_threadedReader->moveToThread(m_fileReadThread);
+  m_threadedReader->setMolecule(m_fileReadMolecule);
+  m_threadedReader->setFileName(fileName);
+
+  // Setup a progress dialog in case file loading is slow
+  m_progressDialog = new QProgressDialog(qobject_cast<QWidget*>(parent()));
+  m_progressDialog->setRange(0, 0);
+  m_progressDialog->setValue(0);
+  m_progressDialog->setMinimumDuration(750);
+  m_progressDialog->setWindowTitle(tr("Reading File"));
+  m_progressDialog->setLabelText(
+    tr("Opening file '%1'\nwith '%2'").arg(fileName).arg(ident));
+  /// @todo Add API to abort file ops
```

Comment:
Wouldn't you just connect the cancel signal with m_fileReadThread => quit() instead?